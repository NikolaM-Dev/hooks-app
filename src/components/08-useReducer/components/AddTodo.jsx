import PropTypes from 'prop-types';

import useForm from 'hooks/useForm';

const AddTodo = ({ handleAddTodo }) => {
  const { values, handleInputChange, reset } = useForm({ description: '' });
  const { description } = values;

  const handleSubmit = (evt) => {
    evt.preventDefault();

    if (description.trim().length <= 1) return;

    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };

    handleAddTodo(newTodo);
    reset();
  };

  return (
    <>
      <h2 className="text-center">Add Todo</h2>
      <hr />

      <form onSubmit={handleSubmit} className="d-grid">
        <input
          autoComplete="off"
          className="form-control"
          name="description"
          onChange={handleInputChange}
          placeholder="Learn ..."
          type="text"
          value={description}
        />
        <button type="submit" className="btn btn-success mt-1">
          Add
        </button>
      </form>
    </>
  );
};

AddTodo.propTypes = {
  handleAddTodo: PropTypes.func.isRequired,
};

export default AddTodo;
