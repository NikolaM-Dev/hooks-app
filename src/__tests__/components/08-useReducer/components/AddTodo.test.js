import '@testing-library/jest-dom';
import { shallow } from 'enzyme';

import AddTodo from 'components/08-useReducer/components/AddTodo';

describe('Tests on <AddTodo /> component', () => {
  const handleAddTodo = jest.fn();

  let wrapper = shallow(<AddTodo handleAddTodo={handleAddTodo} />);

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<AddTodo handleAddTodo={handleAddTodo} />);
  });

  it('should show <AddTodo /> correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("shouldn't call handleAddTodo function", () => {
    const formSubmit = wrapper.find('form').prop('onSubmit');
    formSubmit({ preventDefault() {} });

    expect(handleAddTodo).not.toHaveBeenCalled();
  });

  it('should call handleAddTodo', () => {
    const value = 'Learn React';
    const input = wrapper.find('input');
    input.simulate('change', { target: { value, name: 'description' } });

    const formSubmit = wrapper.find('form').prop('onSubmit');
    formSubmit({ preventDefault() {} });

    expect(handleAddTodo).toHaveBeenCalledTimes(1);
    expect(handleAddTodo).toHaveBeenCalledWith({
      id: expect.any(Number),
      desc: value,
      done: false,
    });
  });

  it('should call reset', () => {
    const value = 'Learn React';
    const input = wrapper.find('input');
    input.simulate('change', { target: { value, name: 'description' } });

    const formSubmit = wrapper.find('form').prop('onSubmit');
    formSubmit({ preventDefault() {} });
    const isEmptyInput = !wrapper.find('input').prop('value');

    expect(isEmptyInput).toBeTruthy();
  });
});
