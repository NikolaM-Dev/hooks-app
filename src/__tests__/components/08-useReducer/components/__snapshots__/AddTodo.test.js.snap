// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <AddTodo /> component should show <AddTodo /> correctly 1`] = `
<Fragment>
  <h2
    className="text-center"
  >
    Add Todo
  </h2>
  <hr />
  <form
    className="d-grid"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Learn ..."
      type="text"
      value=""
    />
    <button
      className="btn btn-success mt-1"
      type="submit"
    >
      Add
    </button>
  </form>
</Fragment>
`;
